feat: add OCI support for Argo CD sources

robinlieb opened the pull request:
## Description

This PR adds the capability to mutate OCI references in Argo CD resources and replaces the given OCI reference to the Zarf registry URL.

This change adds the possibility to use Manifest and Helm Charts in the repoURL in the Application, like explained in the [OCI Docs in Argo CD](https://argo-cd.readthedocs.io/en/stable/user-guide/oci/) with the first two examples. To archive that it also changes the patching of the URL in repository secrets to use the registry instead of the git server if oci is used.

Additionally, added the patching of the repo url for Argo CD [repo-creds](https://argo-cd.readthedocs.io/en/stable/operator-manual/declarative-setup/#repository-credentials) secret type.

## Related Issue

Relates to #3046 

## Checklist before merging

- [X] Test, docs, adr added or updated as needed
- [X] [Contributor Guide Steps](https://github.com/zarf-dev/zarf/blob/main/CONTRIBUTING.md#developer-workflow) followed

robinlieb committed:
feat: add OCI support for Argo CD sources

Signed-off-by: Robin Lieb <34332703+robinlieb@users.noreply.github.com>

brandtkeller reviewed (changes requested):
Made a first pass - spent a healthy amount of time educating myself on argo further. 

Left a nit that I think we can cleanup in a few places. Separately want to consider some additional scope for this change:
1. E2E testing - unit tests get us some good feedback loops on the internal dependency and behaviors but this is largely an external integration. I would love to think about how we test this is operational. 
2. Documentation - right now this can be appending to the existing tutorial. There may be more opportunity for a more granular pass at argocd orchestration docs. 

Two minor requests:
1. Can you comment on an issue if/when you pick it up for work (or at least with the PR) - In doing so we can assign you and help prevent any colliding submissions (as well as remember to track for reviews on the project board. 
2. Add context to the PR description. It took me a fair bit of reading to tell that the associated issue was asking for a feature around oci registries when you actually solved for oci images. The former isn't actually as valid  and you solved the problem correctly here IMO - but I had to hunt down the delta and process. 

Thank you for the continued support and contributions - I very much appreciate it!

robinlieb commented:
@brandtkeller thanks for the initial review. I added changes which addresses all of the mentioned points.
Added two applications in the Argo CD examples showcasing the the use of Helm Charts and manifests from OCI.
Added E2E test on base of this example.
Also refactored the patch logic for Argo CD resources since this contained some duplication.

For the other two points, will definitely keep that in mind for future work.

al-jeyapal commented:
The issue we might have with this is that ArgoCD will try to hit the registry over HTTPS. We were getting this in the Argo logs:

```
Failed to load target state:
 failed to generate manifest for source 1 of 2:
   rpc error: code = Unknown desc = failed to resolve revision "0.0.0":
     cannot get digest for revision 0.0.0:
       Head "https://zarf-docker-registry.zarf.svc.cluster.local:5000/v2/helm-charts/manifests/0.0.0":
         http: server gave HTTP response to HTTPS client
```

Helm allows you to use http via:

```
 helm pull --plain-http oci://...
```

However it doesn't look like Argo exposes an option for that. It exposes a `--insecure-skip-server-verification` but that only allows it to trust unverified certs.

We're currently trying to work around this by putting an internal `caddy` service to act as a HTTPS proxy in front of the zarf registry (on #4327). Still working through the details.

The easier option would be to have a HTTPS endpoint exposed on the zarf registry (just internally as a service not as an ingress/GatewayAPI route), even if it just uses a self-signed certificate. Not sure if that's a route we want to take though.

robinlieb commented:
The [failed pipeline](https://github.com/zarf-dev/zarf/actions/runs/20178960749) Is probably related to the version bump of Argo CD, which throws an error with the gitea.localhost setup in the test. Not sure yet what exactly changed in Argo CD but will investigate further on how to resolve that.

al-jeyapal commented:
Thanks for the work on this @robinlieb. Out of curiosity, how did you work around ArgoCD forcing Helm to use HTTPS when interacting with OCI?  We had to put a HTTPS proxy in place to work around that.

al-jeyapal commented:
> @al-jeyapal setting `insecureOCIForceHttp` to `true` for the repository secret should make that possible. Have you tried that? Had you also the chance to build and test the branch and could provide feedback if that world work on your end?

Thanks @robinlieb I hadn't seen any references to that in the doco, but can see it [in the code](https://github.com/argoproj/argo-cd/blob/c4f3bb8be401e3feb2bf8a85334bd4998eebbbc3/pkg/apis/application/v1alpha1/repository_types.go#L117). Very helpful.

I'm hoping to be rotated back onto the project later this month, so will aim to give it a go when next on.

brandtkeller reviewed (changes requested):
Making my way through each file - only one item stood out as a blocker thus far.

robinlieb commented on src/internal/agent/hooks/argocd-repository.go:
```diff
@@ -107,11 +101,29 @@ func mutateRepositorySecret(ctx context.Context, r *v1.AdmissionRequest, cluster
+	}
+
+	if useMTLS && isOCIURL {
+		patches = append(patches, operations.ReplacePatchOperation("/data/tlsClientCertData", base64.StdEncoding.EncodeToString([]byte(cluster.RegistryClientTLSSecret))))
```
Updated it that both cert and key is used in the Argo CD Repo secret. Updated the test accordingly.

brandtkeller commented on src/internal/agent/hooks/argocd-application.go:
```diff
@@ -118,6 +129,82 @@ func getPatchedRepoURL(ctx context.Context, repoURL string, gs state.GitServerIn
+	return mutateGitURL(ctx, repoURL, gs)
+}
+
+func shouldMutateURL(operation v1.Operation, isOCIURL bool, repoURL, registryAddress, clusterIP string, gs state.GitServerInfo) (bool, bool, error) {
```
Should we unit test and and doccomments for this function? three return values is slightly confusing without any supporting context. I also want to ensure we capture edge cases.

robinlieb replied:
Added docs and unit tests.

brandtkeller commented on src/internal/agent/hooks/argocd-repository.go:
```diff
@@ -107,11 +103,30 @@ func mutateRepositorySecret(ctx context.Context, r *v1.AdmissionRequest, cluster
 	return patches
 }
+
+// Helper for getting eiher git server of registry creds
```
```suggestion
// Helper for getting either git server of registry creds
```

robinlieb replied:
Added change.

brandtkeller reviewed (changes requested):
Minor changes - running this through local testing now to understand the implications.

robinlieb committed:
test: add unit tests for shouldMutateURL function and improve documentation

Signed-off-by: Robin Lieb <34332703+robinlieb@users.noreply.github.com>

brandtkeller commented on examples/argocd/zarf.yaml:
```diff
@@ -8,18 +8,24 @@ components:
           - baseline/values.yaml
+        # Required for 'src/test/external' to validate external git server.
+        values:
+          - sourcePath: ".extraObjects"
```
Deployment of the package will fail if you simply `zarf package create examples/argocd` now unless you provide the required values files on deploy. 

I think the strategy as it stands is sound for adding the testing information at-deploy - but we should include a default values file in the package which has these values specified (empty or defualt) so that the package can still deploy without deploy-time values.

brandtkeller reviewed (changes requested):
I believe tests may be failing related to the comment I made on the argocd package manifest itself. 

Otherwise I'd request one other change (once the argocd package is updated) which would be to run a package create/deploy under the `src/test/proxy/proxy_test.go` in `Test_3_OCIOpsPackage()` such that we continue to test argo for the proxy suite. 

I tested locally - it looks good to me and I can approve if all is good after updates.

robinlieb committed:
feat: add extraObjects and repoServer to values

Signed-off-by: Robin Lieb <34332703+robinlieb@users.noreply.github.com>

robinlieb committed:
build: enable values feature for argocd example

Signed-off-by: Robin Lieb <34332703+robinlieb@users.noreply.github.com>

brandtkeller commented:
Recent PR's touched cluster state - if you can update with main then I'll evaluate any additional testing on airgapped clusters and advocate to get this approved/merged. Thanks for the patience.

robinlieb commented:
> Recent PR's touched cluster state - if you can update with main then I'll evaluate any additional testing on airgapped clusters and advocate to get this approved/merged. Thanks for the patience.

Rebased.

brandtkeller commented on src/internal/agent/hooks/argocd-application.go:
Looking at the new components - I believe we're enabling the orchestration of applications with argocd using only OCI - is that correct?

We made this change recently for an early return when the git server is not configured (for git repo mutations). Now that we have OCI do we need to check for scheme earlier and then determine which service is required?

robinlieb replied:
Good catch, thanks!
Updated that to check for both to have an early return, then check the repoURL for each individual source in the Application, in case there are mixed sources.

brandtkeller reviewed (commented):
Recent change I want to surface - apologies for the churn but this shouldn't be too difficult of a change.

robinlieb committed:
feat: update to new state handling

Signed-off-by: Robin Lieb <34332703+robinlieb@users.noreply.github.com>

zeljkobekcic commented:
Hey, I would really like to use this feature. Can I expect this feature to be in the next release?

brandtkeller commented:
> Hey, I would really like to use this feature. Can I expect this feature to be in the next release?

In review! Next release is tomorrow - so if not that release then the following. 

@robinlieb Can you update with main before I run CI? We had some changes that will likely make them fail if I run now. 

Review nearly complete.

robinlieb commented:
> > Hey, I would really like to use this feature. Can I expect this feature to be in the next release?
> 
> In review! Next release is tomorrow - so if not that release then the following.
> 
> @robinlieb Can you update with main before I run CI? We had some changes that will likely make them fail if I run now.
> 
> Review nearly complete.

@brandtkeller rebased.

brandtkeller commented on src/internal/agent/hooks/argocd-repository.go:
Doing some testing - Do we need to handle registry-only (OCI) for this resource?

robinlieb replied:
Unified the check to conform with applications. Added tests for that.

brandtkeller commented on src/internal/agent/hooks/argocd-appproject.go:
Are there OCI implications that we need to consider here as well?

robinlieb replied:
Unified the check to conform with applications. Added tests for that.

brandtkeller reviewed (changes requested):
I think there are tests implicitly passing with the argo package because we also have the git server configured (required for the package to fully deploy) and the corresponding logic can handle OCI - but when you disable the git server and remove the git-server related elements from the package it does not deploy the applications as expected.

robinlieb committed:
feat: add available service check for project and repo

Signed-off-by: Robin Lieb <34332703+robinlieb@users.noreply.github.com>

brandtkeller commented on src/internal/agent/hooks/argocd-application.go:
```diff
@@ -122,6 +142,95 @@ func getPatchedRepoURL(ctx context.Context, repoURL string, gs state.GitServerIn
+func shouldMutateURL(operation v1.Operation, isOCIURL bool, repoURL, registryAddress, clusterIP string, gs state.GitServerInfo) (shouldMutate bool, isPatchedClusterIP bool, err error) {
+	isCreate := operation == v1.Create
+	isUpdate := operation == v1.Update
+	if isCreate {
```
We removed checks for the create operation #4691, instead reaching for idempotent behaviors. There were edgecases regarding pre-mutated urls on create. I would look to verify if this is required or if the opreation is idempotent regardless.

robinlieb replied:
Removed the operation checks.

brandtkeller commented on src/internal/agent/hooks/common.go:
```diff
@@ -20,6 +21,13 @@ import (
 
+const (
+	// AgentErrTransformGitURL is thrown when the agent fails to make the git url a Zarf compatible url
+	AgentErrTransformGitURL = "unable to transform the git url"
```
nit: if we have these constants then they likeky qualify as good sentinel errors. I won't block on this.

brandtkeller commented on src/internal/agent/hooks/argocd-appproject.go:
```diff
@@ -53,24 +53,32 @@ func mutateAppProject(ctx context.Context, r *v1.AdmissionRequest, cluster *clus
 	for idx, repo := range proj.Spec.SourceRepos {
-		patchedURL, err := getPatchedRepoURL(ctx, repo, s.GitServer)
+		patchedURL, err := getPatchedRepoURL(ctx, repo, registryAddress, clusterIP, s.GitServer, r)
 		// The AppProject can also include source repositories like '*' (as in the default project),
```
Something we should address in the near future is detecting wildcards before the patch operation. This current flow could open up quite a few avenues for unintended consequences on misconfiguration.

brandtkeller commented on src/internal/agent/hooks/argocd-application.go:
```diff
@@ -58,31 +58,46 @@ func NewApplicationMutationHook(ctx context.Context, cluster *cluster.Cluster) o
 		"name", app.Name,
-		"operation", r.Operation,
-		"gitServer", s.GitServer.Address)
+		"operation", r.Operation)
```
Minor nit - we lose the ability to track what the intended mutation was going to try - this has been useful in debugging historically.

robinlieb replied:
Added both, git server and registry.

brandtkeller commented on src/internal/agent/hooks/argocd-application.go:
This check is missing the asymmetry for both git server address and registry address. shouldMutateURL also performs this check in a sense. So I think we just need to resolve how and where to best perform that validation.

brandtkeller commented on src/internal/agent/hooks/argocd-applicationset.go:
```diff
@@ -64,25 +64,40 @@ func mutateApplicationSet(ctx context.Context, r *v1.AdmissionRequest, cluster *
+	l.Info("mutating the ArgoCD ApplicationSet",
 		"name", appSet.Name,
-		"gitServer", s.GitServer.Address)
+		"operation", r.Operation)
```
Same comment for losing address in the logs

robinlieb replied:
Added both, git server and registry.

brandtkeller reviewed (changes requested):
Some of my last comments (hopefully) - only other thing I was going to look at is where we landed for testing with the proxy init model and mTLS.

robinlieb committed:
feat: remove checks for operations

Signed-off-by: Robin Lieb <34332703+robinlieb@users.noreply.github.com>

brandtkeller commented on src/test/e2e/22_git_and_gitops_test.go:
```diff
@@ -192,15 +192,25 @@ func waitFluxPodInfoDeployment(t *testing.T) {
 	// Deploy the argocd example and verify that it works
 	tmpdir := t.TempDir()
-	stdOut, stdErr, err := e2e.Zarf(t, "package", "create", "examples/argocd", "-o", tmpdir, "--skip-sbom")
+	stdOut, stdErr, err := e2e.Zarf(t, "package", "create", "examples/argocd", "-o", tmpdir, "--skip-sbom", "--features=\"values=true\"")
```
For sake of consistency and not introducing any weirdness with cobras stringtostring parser let's be consistent with the features arguments we use elsewhere
```suggestion
	stdOut, stdErr, err := e2e.Zarf(t, "package", "create", "examples/argocd", "-o", tmpdir, "--skip-sbom", "--features", "values=true")
```

brandtkeller commented on src/test/e2e/22_git_and_gitops_test.go:
```diff
@@ -192,15 +192,25 @@ func waitFluxPodInfoDeployment(t *testing.T) {
 	packageName := fmt.Sprintf("zarf-package-argocd-%s.tar.zst", e2e.Arch)
 	path := filepath.Join(tmpdir, packageName)
-	stdOut, stdErr, err = e2e.Zarf(t, "package", "deploy", path, "--components=argocd-apps", "--confirm")
+	stdOut, stdErr, err = e2e.Zarf(t, "package", "deploy", path, "--components=argocd-apps", "--confirm", "--features=\"values=true\"")
```
```suggestion
	stdOut, stdErr, err = e2e.Zarf(t, "package", "deploy", path, "--components=argocd-apps", "--confirm", "--features", "values=true")
```

brandtkeller commented on src/test/external/ext_in_cluster_test.go:
```diff
@@ -107,7 +107,7 @@ func (suite *ExtInClusterTestSuite) Test_0_Mirror() {
 	t := suite.T()
 	tmpdir := t.TempDir()
-	err := exec.CmdWithPrint(zarfBinPath, "package", "create", "../../../examples/argocd", "-o", tmpdir, "--skip-sbom")
+	err := exec.CmdWithPrint(zarfBinPath, "package", "create", "../../../examples/argocd", "-o", tmpdir, "--skip-sbom", "--features=\"values=true\"")
```
```suggestion
	err := exec.CmdWithPrint(zarfBinPath, "package", "create", "../../../examples/argocd", "-o", tmpdir, "--skip-sbom", "--features", "values=true")
```

brandtkeller commented on src/test/external/ext_out_cluster_test.go:
```diff
@@ -110,7 +136,7 @@ func (suite *ExtOutClusterTestSuite) Test_0_Mirror() {
 	t := suite.T()
 	tmpdir := t.TempDir()
-	err := exec.CmdWithPrint(zarfBinPath, "package", "create", "../../../examples/argocd", "-o", tmpdir, "--skip-sbom")
+	err := exec.CmdWithPrint(zarfBinPath, "package", "create", "../../../examples/argocd", "-o", tmpdir, "--skip-sbom", "--features=\"values=true\"")
```
```suggestion
	err := exec.CmdWithPrint(zarfBinPath, "package", "create", "../../../examples/argocd", "-o", tmpdir, "--skip-sbom", "--features", "values=true")
```

brandtkeller reviewed (changes requested):
Minor changes - please rebase with main and I'll kickoff CI and approve/advocate for second review. Appreciate the patience again.

robinlieb committed:
Update src/test/external/ext_out_cluster_test.go

Co-authored-by: Brandt Keller <43887158+brandtkeller@users.noreply.github.com>
Signed-off-by: Robin Lieb <34332703+robinlieb@users.noreply.github.com>

robinlieb committed:
Update src/test/external/ext_in_cluster_test.go

Co-authored-by: Brandt Keller <43887158+brandtkeller@users.noreply.github.com>
Signed-off-by: Robin Lieb <34332703+robinlieb@users.noreply.github.com>

robinlieb committed:
Update src/test/e2e/22_git_and_gitops_test.go

Co-authored-by: Brandt Keller <43887158+brandtkeller@users.noreply.github.com>
Signed-off-by: Robin Lieb <34332703+robinlieb@users.noreply.github.com>

robinlieb committed:
Update src/test/e2e/22_git_and_gitops_test.go

Co-authored-by: Brandt Keller <43887158+brandtkeller@users.noreply.github.com>
Signed-off-by: Robin Lieb <34332703+robinlieb@users.noreply.github.com>

robinlieb commented:
> Minor changes - please rebase with main and I'll kickoff CI and approve/advocate for second review. Appreciate the patience again.

Rebased and proposed changes applied, thanks!

AustinAbro321 reviewed (approved):
LGTM, nice work
